File: android-sdui/src/main/java/com/hsbc/sdui/models/SDUIModels.kt
// SDUIModels.kt
// HSBC SDUI Android Renderer
// Moshi-annotated data classes matching the SDUI JSON schema v1.

package com.hsbc.sdui.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

// ---------------------------------------------------------------------------
// Root payload
// ---------------------------------------------------------------------------

@JsonClass(generateAdapter = true)
data class ScreenPayload(
    @Json(name = "schema_version") val schemaVersion: String,
    @Json(name = "screen")         val screen: String,
    @Json(name = "ttl")            val ttl: Int,          // seconds until stale
    @Json(name = "integrity")      val integrity: String, // SHA-256 of canonical JSON body
    @Json(name = "layout")         val layout: LayoutNode,
    @Json(name = "metadata")       val metadata: ScreenMetadata
)

// ---------------------------------------------------------------------------
// Screen metadata
// ---------------------------------------------------------------------------

@JsonClass(generateAdapter = true)
data class ScreenMetadata(
    @Json(name = "request_id")    val requestId: String,
    @Json(name = "rendered_at")   val renderedAt: String,  // ISO-8601
    @Json(name = "user_id")       val userId: String,
    @Json(name = "segment_id")    val segmentId: String,
    @Json(name = "variant_id")    val variantId: String? = null,
    @Json(name = "experiment_id") val experimentId: String? = null
)

// ---------------------------------------------------------------------------
// Layout node
// ---------------------------------------------------------------------------

@JsonClass(generateAdapter = true)
data class LayoutNode(
    @Json(name = "type")           val type: String,           // "container" | "component"
    @Json(name = "id")             val id: String,
    @Json(name = "component_type") val componentType: String,  // e.g. "promo_banner"
    @Json(name = "props")          val props: Map<String, Any?>,
    @Json(name = "children")       val children: List<LayoutNode>? = null,
    @Json(name = "analytics")      val analytics: AnalyticsConfig? = null,
    @Json(name = "visibility")     val visibility: VisibilityRules? = null
)

// ---------------------------------------------------------------------------
// Analytics configuration
// ---------------------------------------------------------------------------

@JsonClass(generateAdapter = true)
data class AnalyticsConfig(
    @Json(name = "impression_event")  val impressionEvent: String,
    @Json(name = "click_event")       val clickEvent: String? = null,
    @Json(name = "component_id")      val componentId: String,
    @Json(name = "variant_id")        val variantId: String? = null,
    @Json(name = "custom_properties") val customProperties: Map<String, Any?>? = null
)

// ---------------------------------------------------------------------------
// Action definition
// ---------------------------------------------------------------------------

enum class ActionType {
    NAVIGATE,
    DEEP_LINK,
    API_CALL,
    MODAL,
    TRACK,
    SHARE
}

@JsonClass(generateAdapter = true)
data class ActionDefinition(
    @Json(name = "type")        val type: ActionType,
    @Json(name = "destination") val destination: String? = null,
    @Json(name = "params")      val params: Map<String, String>? = null,
    @Json(name = "payload")     val payload: Map<String, Any?>? = null
)

// ---------------------------------------------------------------------------
// Visibility rules
// ---------------------------------------------------------------------------

@JsonClass(generateAdapter = true)
data class VisibilityRules(
    @Json(name = "segment")   val segment: String? = null,
    @Json(name = "platform")  val platform: String? = null,
    @Json(name = "locale")    val locale: String? = null,
    @Json(name = "min_sdui")  val minSdui: String? = null
)

// ---------------------------------------------------------------------------
// Prop accessor helpers (extension functions on Map<String, Any?>)
// ---------------------------------------------------------------------------

fun Map<String, Any?>.getString(key: String): String? = this[key] as? String

fun Map<String, Any?>.getInt(key: String): Int? =
    when (val v = this[key]) {
        is Int    -> v
        is Double -> v.toInt()
        is Long   -> v.toInt()
        else      -> null
    }

fun Map<String, Any?>.getBool(key: String): Boolean? = this[key] as? Boolean

@Suppress("UNCHECKED_CAST")
fun Map<String, Any?>.getMap(key: String): Map<String, Any?>? = this[key] as? Map<String, Any?>

@Suppress("UNCHECKED_CAST")
fun Map<String, Any?>.getList(key: String): List<Any?>? = this[key] as? List<Any?>

/**
 * Attempts to deserialise an `ActionDefinition` stored inline as a prop map.
 * The expected JSON shape is:
 * ```json
 * { "type": "NAVIGATE", "destination": "savings/summer-rate", "params": { ... } }
 * ```
 */
fun Map<String, Any?>.getAction(key: String): ActionDefinition? {
    val raw = getMap(key) ?: return null
    val typeStr = raw.getString("type") ?: return null
    val type = runCatching { ActionType.valueOf(typeStr) }.getOrNull() ?: return null
    val destination = raw.getString("destination")
    @Suppress("UNCHECKED_CAST")
    val params = raw.getMap("params") as? Map<String, String>
    @Suppress("UNCHECKED_CAST")
    val payload = raw.getMap("payload")
    return ActionDefinition(type = type, destination = destination, params = params, payload = payload)
}

File: ios-sdui/Sources/HSBC_SDUI/Models/SDUIModels.swift
// SDUIModels.swift
// HSBC SDUI iOS Renderer
// Models conforming to the SDUI JSON schema v1

import Foundation

// MARK: - Screen Payload

/// Root payload returned by the BFF SDUI endpoint.
public struct ScreenPayload: Codable {
    public let schemaVersion: String
    public let screen: String
    public let ttl: Int                  // seconds until payload is considered stale
    public let integrity: String         // SHA-256 hex of the canonical JSON body
    public let layout: LayoutNode
    public let metadata: ScreenMetadata

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case screen, ttl, integrity, layout, metadata
    }
}

// MARK: - Screen Metadata

public struct ScreenMetadata: Codable {
    public let requestId: String
    public let renderedAt: String        // ISO-8601
    public let userId: String
    public let segmentId: String
    public let variantId: String?
    public let experimentId: String?

    enum CodingKeys: String, CodingKey {
        case requestId   = "request_id"
        case renderedAt  = "rendered_at"
        case userId      = "user_id"
        case segmentId   = "segment_id"
        case variantId   = "variant_id"
        case experimentId = "experiment_id"
    }
}

// MARK: - Layout Node

public enum LayoutNodeType: String, Codable {
    case container
    case component
}

public struct LayoutNode: Codable {
    public let type: LayoutNodeType
    public let id: String
    public let componentType: String     // e.g. "promo_banner", "quick_action_grid"
    public let props: [String: AnyCodable]
    public let children: [LayoutNode]?
    public let analytics: AnalyticsConfig?
    public let visibility: VisibilityRules?

    enum CodingKeys: String, CodingKey {
        case type, id
        case componentType = "component_type"
        case props, children, analytics, visibility
    }
}

// MARK: - Analytics Config

public struct AnalyticsConfig: Codable {
    public let impressionEvent: String
    public let clickEvent: String?
    public let componentId: String
    public let variantId: String?
    public let customProperties: [String: AnyCodable]?

    enum CodingKeys: String, CodingKey {
        case impressionEvent  = "impression_event"
        case clickEvent       = "click_event"
        case componentId      = "component_id"
        case variantId        = "variant_id"
        case customProperties = "custom_properties"
    }
}

// MARK: - Action Definition

public enum ActionType: String, Codable {
    case NAVIGATE
    case DEEP_LINK
    case API_CALL
    case MODAL
    case TRACK
    case SHARE
}

public struct ActionDefinition: Codable {
    public let type: ActionType
    public let destination: String?
    public let params: [String: AnyCodable]?
    public let payload: [String: AnyCodable]?
}

// MARK: - Visibility Rules

public struct VisibilityRules: Codable {
    public let segment: String?
    public let platform: String?
    public let locale: String?
    public let minSdui: String?

    enum CodingKeys: String, CodingKey {
        case segment, platform, locale
        case minSdui = "min_sdui"
    }
}

// MARK: - AnyCodable

/// A type-erased Codable wrapper that can hold arbitrary JSON values
/// (Bool, Int, Double, String, [Any], [String: Any], or null).
public struct AnyCodable: Codable {
    public let value: Any?

    public init(_ value: Any?) {
        self.value = value
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()

        if container.decodeNil() {
            self.value = nil
        } else if let bool = try? container.decode(Bool.self) {
            self.value = bool
        } else if let int = try? container.decode(Int.self) {
            self.value = int
        } else if let double = try? container.decode(Double.self) {
            self.value = double
        } else if let string = try? container.decode(String.self) {
            self.value = string
        } else if let array = try? container.decode([AnyCodable].self) {
            self.value = array.map { $0.value }
        } else if let dict = try? container.decode([String: AnyCodable].self) {
            self.value = dict.mapValues { $0.value }
        } else {
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "AnyCodable: unsupported JSON type"
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch value {
        case nil:
            try container.encodeNil()
        case let bool as Bool:
            try container.encode(bool)
        case let int as Int:
            try container.encode(int)
        case let double as Double:
            try container.encode(double)
        case let string as String:
            try container.encode(string)
        case let array as [Any?]:
            let wrapped = array.map { AnyCodable($0) }
            try container.encode(wrapped)
        case let dict as [String: Any?]:
            let wrapped = dict.mapValues { AnyCodable($0) }
            try container.encode(wrapped)
        default:
            let context = EncodingError.Context(
                codingPath: encoder.codingPath,
                debugDescription: "AnyCodable: value \(String(describing: value)) is not encodable"
            )
            throw EncodingError.invalidValue(value as Any, context)
        }
    }
}

extension AnyCodable: Equatable {
    public static func == (lhs: AnyCodable, rhs: AnyCodable) -> Bool {
        switch (lhs.value, rhs.value) {
        case (nil, nil):                                return true
        case (let l as Bool, let r as Bool):            return l == r
        case (let l as Int, let r as Int):              return l == r
        case (let l as Double, let r as Double):        return l == r
        case (let l as String, let r as String):        return l == r
        default:                                        return false
        }
    }
}

extension AnyCodable: ExpressibleByNilLiteral {
    public init(nilLiteral: ()) { self.value = nil }
}

extension AnyCodable: ExpressibleByStringLiteral {
    public init(stringLiteral value: String) { self.value = value }
}

extension AnyCodable: ExpressibleByIntegerLiteral {
    public init(integerLiteral value: Int) { self.value = value }
}

extension AnyCodable: ExpressibleByBooleanLiteral {
    public init(booleanLiteral value: Bool) { self.value = value }
}

// MARK: - Convenience helpers on AnyCodable

public extension AnyCodable {
    var stringValue: String? { value as? String }
    var intValue: Int? { value as? Int }
    var doubleValue: Double? { value as? Double }
    var boolValue: Bool? { value as? Bool }
    var dictValue: [String: Any?]? { value as? [String: Any?] }
    var arrayValue: [Any?]? { value as? [Any?] }
}
